### PR TITLE
 Add ability to perform prediction on CPU

### LINKED ISSUE #41: Add support for CPU-only inference
- [x] Merge https://github.com/dmlc/treelite/pull/274
- [x] Merge https://github.com/rapidsai/cuml/pull/3855
- [x] Merge #23
- [x] Update configuration to allow selection of CPU-only mode
- [x] Update batching to optionally prefer host arrays
- [x] Use GTIL prediction call when in CPU-only mode

### PR DESCRIPTION
**Pull request by hcho3**
It is now possible to run prediction on CPU. Just set
```
instance_group [{ kind: KIND_CPU }]
```
in the `config.pbtxt`.

Closes #41

Replaces #63

### CODE REVIEW

File: src/gtil_utils.cu [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+  std::size_t num_class = 1;
+  auto tl_model = static_cast<treelite::Model*>(model_state.treelite_handle);
+  num_class = tl_model->task_param.num_class;
+  if (!predict_proba && model_state.tl_params.output_class && num_class > 1) {
+    std::strcpy(tl_model->param.pred_transform, "max_index");
+  }
+
+  std::size_t out_result_size;
+  int gtil_result = TreeliteGTILPredict(model_state.treelite_handle, data.data(),
+              data.shape()[0], preds.data(), 1, &out_result_size);
```
Comment (hcho3) [inline src/gtil_utils.cu:51]: For a multi-class classifier with `predict_proba=False`, GTIL requires the buffer of size `num_row * num_class` to store the raw scores. So in this case `out_result_size == num_row < num_row * num_class`

**Comment by hcho3** [comment]
This PR is now ready for review. I tested it locally and all tests passed.

File: src/model_instance_state.cu [hunk 2]
```diff
@@ -33,54 +37,81 @@ ModelInstanceState::Create(
...
+        }
+      }())
 {
   model_state_->LoadModel(ArtifactFilename(), Kind(), DeviceId());
-  ML::fil::from_treelite(
-      *handle, &fil_forest, model_state_->treelite_handle,
-      &(model_state_->tl_params));
+
+  if (Kind() == TRITONSERVER_INSTANCEGROUPKIND_GPU) {
+    LOG_MESSAGE(TRITONSERVER_LOG_INFO, "Using GPU for inference");
```
Comment (wphicks) [inline src/model_instance_state.cu:96]: Let's make this and the next log message slightly more specific to clarify that we're using GPU for inference with a particular model or at least that this applies to only the FIL backend

**Review by wphicks — CHANGES_REQUESTED** [review]

**Comment by wphicks** [comment]
Just created #88 as a follow-on item. No need for us to fix it in this PR, but we should clean up the duplication across CPU/GPU tests both for code cleanliness and CI efficiency.

File: src/gtil_utils.cu [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  // avoid out-of-bounds access.
+  bool use_temp_buffer = false;
+  std::vector<float> temp_buffer;
+  if (!predict_proba && model_state.tl_params.output_class && num_class > 1) {
+    std::strcpy(tl_model->param.pred_transform, "max_index");
+    use_temp_buffer = true;
+  }
+
+  float* out_preds = preds.data();
+  if (use_temp_buffer) {
```
[the remark this answers is not in the document]
Comment (hcho3) [reply src/gtil_utils.cu:59]: > when checking whether a copy is needed, we can either check if temp_buffer has non-zero size or (even better) wrap temp_buffer in std::optional and only construct it if we need it.

I'm not convinced that implicitly encoding the need for copy in the size of `temp_buffer` is necessarily better than explicitly using a flag.

[the remark this answers is not in the document]
Comment (hcho3) [reply src/gtil_utils.cu:59]: On a second thought, we can wrap `temp_buffer` in `std::optional` and check for its existence to replace `use_temp_buffer` flag. Let me make the change now.

**Review by wphicks — APPROVED** [review]
Looks great to me! Let's hold off on merging until @divyegala and @GuanLuo have weighed in, but I think we're looking very nice.

File: src/triton_fil/triton_tensor.cuh [hunk 4]
```diff
@@ -128,23 +164,33 @@ class TritonTensor {
...
-            raft::copy(ptr_d, other.buffer, other.size(), stream_);
-          }
-          catch (const raft::cuda_error& err) {
-            throw TritonException(
-                TRITONSERVER_errorcode_enum::TRITONSERVER_ERROR_INTERNAL,
-                err.what());
+        target_memory_{other.target_memory_}, size_bytes_{other.size_bytes_},
+        is_owner_{true}, stream_{other.stream_}, buffer{[&] {
+          non_const_T* ptr_d;
+          if (other.target_memory_ == TRITONSERVER_MEMORY_GPU) {
```
Comment (divyegala) [inline src/triton_fil/triton_tensor.cuh:180]: I see these kinds of copies being repeated a few times. Can this be pulled out in a utility function with the try-catch and letting the function maybe even handle the host copy part?

**Review by divyegala — COMMENTED** [review]
Some small suggestions, the additions look pretty nice!

**Review by GuanLuo — APPROVED** [review]
LGTM

File: src/gtil_utils.cu [hunk 5]
```diff
@@ -0,0 +1,88 @@
...
+#include <triton_fil/model_instance_state.cuh>
+#include <triton_fil/triton_tensor.cuh>
+
+namespace triton { namespace backend { namespace fil {
+void
+gtil_predict(
+    ModelInstanceState& instance_state, TritonTensor<const float>& data,
+    TritonTensor<float>& preds, bool predict_proba)
+{
+  auto& model_state = [&]() -> ModelState& {
```
[the remark this answers is not in the document]
Comment (hcho3) [reply src/gtil_utils.cu]: `instance_state` is captured by reference. I made the capture explicit.

File: src/triton_fil/triton_tensor.cuh [hunk 6]
```diff
@@ -55,38 +57,70 @@ class TritonTensor {
  public:
   TritonTensor()
       : name_{}, shape_{}, dtype_{TRITONSERVER_TYPE_FP32}, size_bytes_{0},
-        is_owner_{false}, stream_{0}, buffer{nullptr}, final_buffers{}
+        target_memory_{TRITONSERVER_MEMORY_GPU}, is_owner_{false}, stream_{0},
+        buffer{nullptr}, final_buffers{}
   {
   }
 
   template <typename U = T>
```
[the remark this answers is not in the document]
Comment (hcho3) [reply src/triton_fil/triton_tensor.cuh:65]: Let me remove it.

File: src/triton_fil/triton_tensor.cuh (hunk 4 above)
Comment (hcho3) [reply src/triton_fil/triton_tensor.cuh:169]: I created a new utility function `memory_copy`.

**Comment by wphicks** [comment]
CI failure just needs latest changes from `main` merged in. Otherwise, the new changes look good! I particularly like the new utility function.

**Review by divyegala — APPROVED** [review]
LGTM, thank you!

**Pull request merged**